Fixes parsing of GC entries in elasticsearch server log

ycombinator opened the pull request:
Resolves #9513.

This PR:
* removes the incorrectly-parsed `gc_overhead` field. Turns out what we were parsing was actually an insignificant sequential number, not GC overhead,
* parses out a new  `gc.collection_duration` field, e.g. `1.2s`, which is the time spent performing GC, and
* parses out a new `gc.observation_duration` field, e.g. `1.8s`, which is the overall time over which GC collection was performed

It also splits up the long grok expression in the ingest pipeline into smaller patterns and references those patterns, hopefully for easier readability.

kaiyan-sheng commented:
@ycombinator Did you try to run `mage fmt update`?

ruflin commented:
@ycombinator Could you rebase on master. The failing CI should be fixed there.

ycombinator commented:
@ruflin Rebased on `master` now.

ycombinator commented:
@radoondas has suggested a few improvements so I'm taking this PR out of `review` for now. I will put it back into `review` after I've incorporated his improvements.

ycombinator commented:
Thanks to @radoondas's work, the ingest pipeline in this PR now includes a script processor to parse the GC durations into meaningful numeric millisecond values. It also further breaks down the grok expressions to make them even more readable.

tsouza commented:
@ycombinator you've described the metrics as `Old GC`. According to https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/monitor/jvm/JvmGcMonitorService.java#L458 it's not only `Old GC`. It's the overhead of all the GCs.

Also, I am not sure if collecting the seq number is of any usage.

ycombinator commented:
Thanks @tsouza. Fixing both those things...

ycombinator commented:
I've incorporated the latest feedback into this PR. It is ready for review again. Thanks!

ycombinator commented:
> I wonder if an improvement can be made later on on the ES side to always log in the same unit?

@ruflin Added https://github.com/elastic/elasticsearch/issues/32850#issuecomment-449012625.

I've also spoken with @jakelandis about the possibility of adding a `duration` Ingest processor, similar to the `bytes` one we have today for performing the necessary parsing and conversion from string durations to numeric durations.

jsoriano commented on filebeat/module/elasticsearch/server/ingest/pipeline.json:
```diff
@@ -20,13 +20,24 @@
+        {
+            "script": {
+                "lang": "painless",
+                "source": "if (ctx.elasticsearch.server.gc != null && ctx.elasticsearch.server.gc.observation_duration != null) { if (ctx.elasticsearch.server.gc.observation_duration.unit == 's') { ctx.elasticsearch.server.gc.observation_duration.time_in_millis = ctx.elasticsearch.server.gc.observation_duration.time * 1000;}if (ctx.elasticsearch.server.gc.observation_duration.unit == 'ms') { ctx. ...
```
Can literals here also trigger recompilations as in https://github.com/elastic/beats/issues/9600#issuecomment-448255335 ?

ycombinator replied:
Wow, so odd! In any case, I'm okay parameterizing the literals, just in case. Thanks for pointing me to this.

jsoriano replied:
Do you know if there is a way to know if a script is being recompiled when data is ingested?

ycombinator commented:
@jsoriano Thanks for your review. I've addressed all your feedback now. Ready for your 👀 again!

Additionally I've also shortened the painless expression a bit by using the [null-safe operator](https://www.elastic.co/guide/en/elasticsearch/painless/master/painless-operators-reference.html#null-safe-operator).

ycombinator commented:
jenkins, test this

jpountz commented on filebeat/module/elasticsearch/server/ingest/pipeline.json:
```diff
@@ -20,13 +20,24 @@
+        {
+            "script": {
+                "lang": "painless",
+                "source": "if (ctx.elasticsearch.server.gc != null && ctx.elasticsearch.server.gc.observation_duration != null) { if (ctx.elasticsearch.server.gc.observation_duration.unit == 's') { ctx.elasticsearch.server.gc.observation_duration.time_in_millis = ctx.elasticsearch.server.gc.observation_duration.time * 1000;}if (ctx.elasticsearch.server.gc.observation_duration.unit == 'ms') { ctx. ...
```
:)
